File: android/src/main/java/com/khipu/capacitor/KhipuResultReader.java
package com.khipu.capacitor;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.khipu.client.KhipuEvent;
import com.khipu.client.KhipuResult;
import java.io.Serializable;
import org.json.JSONObject;

/**
 * Turns the activity's payload into the object handed back to JS, or null when there is
 * no usable payload.
 *
 * Takes the extra rather than the Intent so it can be unit tested without Android, and
 * so it cannot see the activity's result code. That is deliberate: both of the SDK's
 * exits carry a complete KhipuResult — the normal one at KhipuActivity.kt:320 and the
 * destroyed-too-long abort at :119, which reports USER_CANCELED. Branching on the code
 * would make the same outcome, the user abandoning the payment, arrive in two different
 * shapes depending on an invisible timing detail.
 *
 * The object is built key by key rather than through KhipuResult.asJson(): that helper
 * puts the field names out of reach of the vocabulary guard. (It is otherwise a fine
 * encoding — since 2.28.4 it builds its Gson with serializeNulls, so it emits the same
 * present-as-null shape this reader now produces.)
 */
final class KhipuResultReader {

    private KhipuResultReader() {}

    static JSObject read(Serializable extra) {
        if (!(extra instanceof KhipuResult)) {
            return null;
        }

        KhipuResult source = (KhipuResult) extra;
        JSObject result = new JSObject();
        put(result, "operationId", source.getOperationId());
        put(result, "exitTitle", source.getExitTitle());
        put(result, "exitMessage", source.getExitMessage());
        put(result, "result", source.getResult());
        put(result, "exitUrl", source.getExitUrl());
        put(result, "continueUrl", source.getContinueUrl());
        put(result, "failureReason", source.getFailureReason());
        put(result, "events", events(source.getEvents()));
        return result;
    }

    /**
     * Every key goes through here, in one call shape, for two reasons.
     *
     * A null value is written as JSONObject.NULL, not skipped, so the key set never
     * varies with the outcome of the payment: all eight are always present and the
     * three optional ones carry an explicit null. JSONObject.NULL rather than a bare
     * null because JSONObject.put REMOVES the key when handed one — that is exactly how
     * this side came to omit them in the first place, and it is silent.
     *
     * And one uniform shape is what `check-option-keys.mjs` greps for. A second
     * spelling would hide those keys from the guard — which is exactly how the old
     * `asJson()` call kept the whole result out of its reach.
     */
    private static void put(JSObject target, String key, Object value) {
        target.put(key, value == null ? JSONObject.NULL : value);
    }

    private static JSArray events(KhipuEvent[] events) {
        JSArray array = new JSArray();
        if (events == null) {
            return array;
        }
        for (KhipuEvent event : events) {
            JSObject item = new JSObject();
            item.put("name", event.getName());
            item.put("timestamp", event.getTimestamp());
            item.put("type", event.getType());
            array.put(item);
        }
        return array;
    }
}

File: ios/Sources/KhipuPlugin/KhipuResultReader.swift
import Foundation
import KhipuClientIOS

/// Turns the SDK's result into the object handed back to JS.
///
/// **The key set never varies with the outcome of the payment.** All eight keys are
/// always present; the three the SDK declares optional — `exitUrl`, `continueUrl` and
/// `failureReason` — carry an explicit `null` when it has no value for them. That is
/// what `cordova-khipu` and `react-native-khipu` do, and what the Android SDK's own
/// `asJson()` does (it builds its `Gson` with `serializeNulls`).
///
/// This file has now held both encodings, so the history is worth keeping. It began as
/// `result.exitUrl as Any` inside the launch closure: an empty Swift optional cast to
/// `Any` is not nil, it is an `Any` wrapping the empty optional, so the key was written
/// and the bridge serialised it as null — the right wire format by accident, while
/// Android omitted the same keys. `8.0.0` aligned them on omitting, reasoning that
/// `KhipuResult` declared `string | undefined` and null did not satisfy it. That
/// reasoning was weaker than it looked: `tsc --strict` flags neither `=== null` against
/// `string | undefined` nor `=== undefined` against `string | null`, so the declaration
/// was as unreviewed an artefact as the cast, and treating it as ground truth begged the
/// question. The family had meanwhile settled on present-as-null, so that is where this
/// lands — this time as a decision rather than a property of a cast.
///
/// Extracted from the closure so it can be unit tested at all, mirroring
/// `KhipuResultReader.java` on the other side.
enum KhipuResultReader {

    static func read(_ source: KhipuResult) -> [String: Any] {
        var result: [String: Any] = [:]
        put(&result, "operationId", source.operationId)
        put(&result, "exitTitle", source.exitTitle)
        put(&result, "exitMessage", source.exitMessage)
        put(&result, "result", source.result)
        put(&result, "exitUrl", source.exitUrl)
        put(&result, "continueUrl", source.continueUrl)
        put(&result, "failureReason", source.failureReason)
        put(&result, "events", events(source.events))
        return result
    }

    /// Every key goes through here, in one call shape, for the same two reasons as the
    /// Java counterpart.
    ///
    /// An empty value is written as `NSNull`, not skipped: assigning nil to a
    /// `[String: Any]` subscript *removes* the key, which is the very behaviour this no
    /// longer wants. `NSNull` is what Capacitor's bridge serialises to JSON `null`.
    ///
    /// And one uniform shape is what `check-option-keys.mjs` greps for; a second
    /// spelling would hide those keys from the guard.
    ///
    /// The parameter is generic rather than `Any?` deliberately. Passing a `String?`
    /// into an `Any?` parameter re-opens the boxing question that produced the original
    /// defect; with `T?` there is no `Any` for an empty optional to hide inside.
    private static func put<T>(_ target: inout [String: Any], _ key: String, _ value: T?) {
        if let value {
            target[key] = value
        } else {
            target[key] = NSNull()
        }
    }

    private static func events(_ events: [KhipuEvent]) -> [[String: String]] {
        events.map { ["name": $0.name, "timestamp": $0.timestamp, "type": $0.type] }
    }
}
